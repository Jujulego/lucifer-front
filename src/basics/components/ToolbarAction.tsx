import React, { ElementType } from 'react';

import {
  ExtendButtonBase,
  ExtendButtonBaseTypeMap,
  IconButton, IconButtonClassKey, IconButtonTypeMap,
  Tooltip, TooltipProps
} from '@material-ui/core';

import { OverrideProps } from '@material-ui/core/OverridableComponent';

// Types
export type ToolbarActionClassKey = Exclude<IconButtonClassKey, 'colorPrimary' | 'colorSecondary'>;
export type ToolbarActionTypeMap<
  P = {},
  D extends ElementType = 'button'
> = ExtendButtonBaseTypeMap<{
  props: P & Omit<IconButtonTypeMap<P, D>['props'], 'color'> & {
    tooltip: string,
    tooltipProps?: Omit<TooltipProps, 'title'>
  };
  defaultComponent: D;
  classKey: ToolbarActionClassKey;
}>;

export type ToolbarActionProps<
  D extends ElementType = ToolbarActionTypeMap['defaultComponent'],
  P = {}
> = OverrideProps<ToolbarActionTypeMap<P, D>, D>;

// Component
const ToolbarAction: ExtendButtonBase<ToolbarActionTypeMap> = <D extends ElementType> (props: { component?: D } & ToolbarActionProps<D>) => {
  // Props
  const {
    children, disabled,
    tooltip, tooltipProps,
    ...button
  } = props;

  // Render
  const btn = (
    <IconButton {...button} disabled={disabled} color="inherit">
      { children }
    </IconButton>
  );

  if (disabled) return btn;
  return (
    <Tooltip {...tooltipProps} title={tooltip}>
      { btn }
    </Tooltip>
  );
};

export default ToolbarAction;
