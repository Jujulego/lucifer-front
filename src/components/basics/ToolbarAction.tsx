import React, { ElementType } from 'react';

import {
  IconButton, ExtendButtonBase, IconButtonClassKey, IconButtonTypeMap,
  Tooltip, TooltipProps
} from '@material-ui/core';
import { OverrideProps } from '@material-ui/core/OverridableComponent';

// Types
export type ToolbarActionClassKey = Exclude<IconButtonClassKey, 'colorPrimary' | 'colorSecondary'>;

export type ToolbarActionTypeMap<
  P = {},
  D extends ElementType = IconButtonTypeMap['defaultComponent']
> = {
  props: Omit<IconButtonTypeMap<P, D>['props'], 'color'> & {
    tooltip: string,
    tooltipProps?: Omit<TooltipProps, 'title'>
  },
  defaultComponent: D,
  classKey: ToolbarActionClassKey
};

export type ToolbarActionProps<
  D extends ElementType = ToolbarActionTypeMap['defaultComponent'],
  P = {}
> = OverrideProps<ToolbarActionTypeMap<P, D>, D>;

// Component
const ToolbarAction: ExtendButtonBase<ToolbarActionTypeMap> = (props: ToolbarActionProps) => {
  // Props
  const { children, tooltip, tooltipProps, ...button } = props;

  // Render
  return (
    <Tooltip {...tooltipProps} title={tooltip}>
      <IconButton {...button} color="inherit">
        {children}
      </IconButton>
    </Tooltip>
  );
};

export default ToolbarAction;