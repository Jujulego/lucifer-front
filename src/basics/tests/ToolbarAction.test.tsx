import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import ToolbarAction from '../components/ToolbarAction';

// Setup
let mount: ReturnType<typeof createMount>;
let shallow: ReturnType<typeof createShallow>;

beforeAll(() => {
  mount = createMount();
  shallow = createShallow();
});

afterAll(() => {
  mount.cleanUp();
});

// Tests
it('should render correctly', () => {
  // Render
  const wrapper = shallow(
    <ToolbarAction tooltip="Test" />
  );

  // Check elements
  expect(wrapper).toMatchSnapshot();
});

it('should react on click', () => {
  const spy = jest.fn();

  // Render
  const wrapper = mount(
    <ToolbarAction tooltip="Test" onClick={spy} />
  );

  // Get button
  const button = wrapper.find('button');
  expect(button).toHaveLength(1);

  // Test event
  button.simulate('click');

  expect(spy).toBeCalledTimes(1);
});

it('should have no tooltip when disabled', () => {
  // Render
  const wrapper = shallow(
    <ToolbarAction tooltip="Test" disabled />
  );

  // Check elements
  expect(wrapper).toMatchSnapshot();
});

it('should not react when disabled', () => {
  const spy = jest.fn();

  // Render
  const wrapper = mount(
    <ToolbarAction tooltip="Test" onClick={spy} disabled />
  );

  // Get button
  const button = wrapper.find('button');
  expect(button).toHaveLength(1);
  expect(button.prop('disabled')).toBeTruthy();

  // Test event
  button.simulate('click');
  expect(spy).not.toBeCalled();
});
