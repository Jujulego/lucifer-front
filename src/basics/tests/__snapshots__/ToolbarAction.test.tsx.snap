// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have no tooltip when disabled 1`] = `
<WithStyles(ForwardRef(IconButton))
  color="inherit"
  disabled={true}
/>
`;

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(Tooltip))
  title="Test"
>
  <WithStyles(ForwardRef(IconButton))
    color="inherit"
  />
</WithStyles(ForwardRef(Tooltip))>
`;
